lib/charms/opensearch/v0/opensearch_backups.py

```diff
@@ -0,0 +1,538 @@
+
+
+class OpenSearchBackupError(OpenSearchError):
+    """Exception thrown when an opensearch backup-related action fails."""
+
+
+class OpenSearchListBackupError(OpenSearchBackupError):
+    """Exception thrown when internal list backups call fails."""
+
+
+class BackupServiceState(BaseStrEnum):
+    """Enum for the states possible once plugin is enabled."""
+
+    SUCCESS = "success"
+    RESPONSE_FAILED_NETWORK = "response failed: network error"
+    REPO_NOT_CREATED = "repository not created"
+    REPO_NOT_CREATED_ALREADY_EXISTS = "repo not created as it already exists"
+    REPO_CREATION_ERR = "Failed creating repository"
+    REPO_ERR_UNKNOWN = "Repository exception: unknown"
+    REPO_MISSING = "repository is missing from request"
+    REPO_S3_UNREACHABLE = "repository s3 is unreachable"
+    ILLEGAL_ARGUMENT = "request contained wrong argument"
+    SNAPSHOT_MISSING = "snapshot not found"
+    SNAPSHOT_RESTORE_ERROR = "restore of snapshot failed"
+    SNAPSHOT_IN_PROGRESS = "snapshot in progress"
+    SNAPSHOT_PARTIALLY_TAKEN = "snapshot partial: at least one shard missing"
+    SNAPSHOT_INCOMPATIBILITY = "snapshot failed: incompatibility issues"
+    SNAPSHOT_FAILED_UNKNOWN = "snapshot failed for unknown reason"
+
+
+class OpenSearchBackup(Object):
+    """Implements backup relation and API management."""
+
+    def __init__(self, charm: Object):
+        """Manager of OpenSearch backup relations."""
+        super().__init__(charm, S3_RELATION)
+        self.charm = charm
+        # s3 relation handles the config options for s3 backups
+        self.s3_client = S3Requirer(self.charm, S3_RELATION)
+        self.framework.observe(self.charm.on[S3_RELATION].relation_broken, self._on_s3_broken)
+        self.framework.observe(
+            self.s3_client.on.credentials_changed, self._on_s3_credentials_changed
+        )
+        self.framework.observe(self.charm.on.create_backup_action, self._on_create_backup_action)
+        self.framework.observe(self.charm.on.list_backups_action, self._on_list_backups_action)
+
+    @property
+    def _plugin_status(self):
+        return self.charm.plugin_manager.get_plugin_status(OpenSearchBackupPlugin)
+
+    def _on_list_backups_action(self, event: ActionEvent) -> None:
+        """Returns the list of available backups to the user."""
+        if not self._check_action_can_run(event):
+            event.fail("Failed: backup service is not configured yet")
+            return
+        backups = {}
+        try:
+            backups = self._fetch_snapshots()
+            event.set_results({"snapshots": (json.dumps(backups)).replace("_", "-")})
+        except OpenSearchError as e:
+            event.fail(
+                f"List backups action failed - {str(e)} - check the application logs for the full stack trace."
+            )
+
+    def _on_create_backup_action(self, event: ActionEvent) -> None:
+        """Creates a backup from the current cluster."""
+        if not self.charm.unit.is_leader():
```

Comment:
just Q: why only leader? it is convenient if you can run a backup on particular (e.g. less busy) unit